lib/semantics/check-declarations.cc

```diff
@@ -457,6 +469,137 @@ static bool ConflictsWithIntrinsicAssignment(const Procedure &proc) {
       IsDefinedAssignment(lhs.type(), lhs.Rank(), rhs.type(), rhs.Rank());
 }
 
+static bool ConflictsWithIntrinsicOperator(
+    const GenericKind &kind, const Procedure &proc) {
+  auto arg0{std::get<DummyDataObject>(proc.dummyArguments[0].u).type};
+  auto type0{arg0.type()};
+  if (proc.dummyArguments.size() == 1) {  // unary
+    return std::visit(
+        common::visitors{
+            [&](common::NumericOperator) { return IsIntrinsicNumeric(type0); },
+            [&](common::LogicalOperator) { return IsIntrinsicLogical(type0); },
+            [](const auto &) -> bool { DIE("bad generic kind"); },
+        },
+        kind.u);
+  } else {  // binary
+    int rank0{arg0.Rank()};
+    auto arg1{std::get<DummyDataObject>(proc.dummyArguments[1].u).type};
+    auto type1{arg1.type()};
+    int rank1{arg1.Rank()};
+    return std::visit(
+        common::visitors{
+            [&](common::NumericOperator) {
+              return IsIntrinsicNumeric(type0, rank0, type1, rank1);
+            },
+            [&](common::LogicalOperator) {
+              return IsIntrinsicLogical(type0, rank0, type1, rank1);
+            },
+            [&](common::RelationalOperator opr) {
+              return IsIntrinsicRelational(opr, type0, rank0, type1, rank1);
+            },
+            [&](GenericKind::OtherKind x) {
+              CHECK(x == GenericKind::OtherKind::Concat);
+              return IsIntrinsicConcat(type0, rank0, type1, rank1);
+            },
+            [](const auto &) -> bool { DIE("bad generic kind"); },
+        },
+        kind.u);
+  }
+}
+
+// Check if this procedure can be used for defined operators (see 15.4.3.4.2).
+bool CheckHelper::CheckDefinedOperator(const SourceName &opName,
+    const GenericKind &kind, const Symbol &specific, const Procedure &proc) {
+  std::optional<parser::MessageFixedText> msg;
+  if (!proc.functionResult.has_value()) {
+    msg = "%s procedure '%s' must be a function"_err_en_US;
+  } else if (proc.functionResult->IsAssumedLengthCharacter()) {
+    msg = "%s function '%s' may not have assumed-length CHARACTER(*)"
+          " result"_err_en_US;
+  } else if (!CheckNumberOfArgs(msg, kind, proc.dummyArguments.size())) {
+    // error is in msg
+  } else if (!CheckDefinedOperatorArg(opName, specific, proc, 0) |
+      !CheckDefinedOperatorArg(opName, specific, proc, 1)) {
+    return false;  // error was reported
+  } else if (ConflictsWithIntrinsicOperator(kind, proc)) {
+    msg = "%s function '%s' conflicts with intrinsic operator"_err_en_US;
+  } else {
+    return true;  // OK
+  }
+  SayWithDeclaration(specific, std::move(msg.value()),
+      parser::ToUpperCaseLetters(opName.ToString()), specific.name());
+  return false;
+}
+
+// If the number of arguments is wrong for this intrinsic operator, return
```

Comment:
Returning just an optional message (whose presence signifies an error) would be a cleaner interface.

Reply:
OK, I've made that change.